<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Introduction to Bioinformatics workflows with Nextflow and nf-core: Simple RNA-Seq pipeline</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../11-Simple_Rna-Seq_pipeline.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Introduction to Bioinformatics workflows with Nextflow and nf-core
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Introduction to Bioinformatics workflows with Nextflow and nf-core
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="learner_profiles.html">Learner Profiles</a></li><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Groovy syntax</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to Bioinformatics workflows with Nextflow and nf-core
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 79%" class="percentage">
    79%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 79%" aria-valuenow="79" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../11-Simple_Rna-Seq_pipeline.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-getting-started-with-nextflow.html">1. Getting Started with Nextflow</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-workflow_parameters.html">2. Workflow parameterisation</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-channels.html">3. Channels</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-processes-part1.html">4. Processes</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-processes-part2.html">5. Processes Part 2</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-operators.html">6. Operators</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-workflow.html">7. Workflow</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-reporting.html">8. Reporting</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-workflow_checkpoint_caching.html">9. Workflow caching and checkpointing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-configuration.html">10. Nextflow configuration</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        11. Simple RNA-Seq pipeline
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#define-the-pipeline-parameters">Define the pipeline parameters</a></li>
<li><a href="#create-a-transcriptome-index-file">Create a transcriptome index file</a></li>
<li><a href="#collect-read-files-by-pairs">Collect read files by pairs</a></li>
<li><a href="#perform-expression-quantification">Perform expression quantification</a></li>
<li><a href="#add-a-publishdir-directive">Add a publishDir directive</a></li>
<li><a href="#quality-control">Quality control</a></li>
<li><a href="#multiqc-report">MultiQC report</a></li>
<li><a href="#handle-completion-event">Handle completion event</a></li>
<li><a href="#metrics-and-reports">Metrics and reports</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush13">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading13">
        <a href="12-nfcore.html">12. Deploying nf-core pipelines</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <li><a href="learner_profiles.html">Learner Profiles</a></li><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Groovy syntax</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/10-configuration.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/12-nfcore.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/10-configuration.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Nextflow
        </a>
        <a class="chapter-link float-end" href="../instructor/12-nfcore.html" rel="next">
          Next: Deploying nf-core...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Simple RNA-Seq pipeline</h1>
        <p>Last updated on 2025-01-06 |

        <a href="https://github.com/carpentries-incubator/workflows-nextflow/edit/main/episodes/11-Simple_Rna-Seq_pipeline.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can I create a Nextflow pipeline from a series of unix commands
and input data?</li>
<li>How do I log my pipelines parameters?</li>
<li>How can I manage my pipeline software requirements?</li>
<li>How do I know when my pipeline has finished?</li>
<li>How do I see how much resources my pipeline has used?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Create a simple RNA-Seq pipeline.</li>
<li>Use the <code>log.info</code> function to print all the pipeline
parameters.</li>
<li>Print a confirmation message when the pipeline completes.</li>
<li>Use a conda <code>environment.yml</code> file to install the
pipeline’s software requirement.</li>
<li>Produce an execution report and generate run metrics from a pipeline
run.</li>
</ul></div>
</div>
</div>
</div>
</div>
<p>We’re now set to develop a multi-step pipeline using Nextflow, for
analyzing and performing quality control on our yeast RNA-Seq
experiment.</p>
<p>In this RNA-Seq pipeline, we’ll undertake the following steps to
thoroughly analyze gene expression data:</p>
<ol style="list-style-type: decimal"><li>
<strong>Quality Control with FastQC</strong>: <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
assesses the quality of the data by generating reports that highlight
any potential issues, such as low-quality sequences or contamination.
FastQC’s output includes an HTML report and a directory containing
detailed analyses, which are essential for evaluating the integrity of
the sequencing data.</li>
</ol><div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> mkdir fastqc_<span class="op">&lt;</span>sample_id<span class="op">&gt;</span>_logs</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="at">-o</span> fastqc_<span class="op">&lt;</span>sample_id<span class="op">&gt;</span>_logs <span class="at">-f</span> fastq <span class="at">-q</span> <span class="op">&lt;</span>reads<span class="op">&gt;</span></span></code></pre>
</div>
<ol start="2" style="list-style-type: decimal"><li>
<strong>Transcriptome Indexing with Salmon</strong>: <a href="https://salmon.readthedocs.io/en/latest/" class="external-link">Salmon</a> is a tool to
quantify transcript expression in RNA-seq data. The first step in the
process is for Salmon to create an index of the transcriptome. This step
involves processing a reference transcriptome, which allows for
efficient and accurate mapping and quantification of RNA-Seq reads.</li>
</ol><div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">$</span> salmon index <span class="at">--threads</span> <span class="op">&lt;</span>cpus<span class="op">&gt;</span> -t <span class="op">&lt;</span>transcriptome file<span class="op">&gt;</span> -i index</span></code></pre>
</div>
<ol start="3" style="list-style-type: decimal"><li>
<strong>Quantification with Salmon</strong>: After indexing, Salmon
is used for the quantification step. In this step, Salmon maps the reads
to the transcriptome index and quantifies the transcript abundances.
This process is crucial for determining the expression levels of genes
in the sample.</li>
</ol><div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> salmon quant <span class="at">--threads</span> <span class="op">&lt;</span>cpus<span class="op">&gt;</span> --libType=U <span class="at">-i</span> <span class="op">&lt;</span>index<span class="op">&gt;</span> -1 <span class="op">&lt;</span>read1<span class="op">&gt;</span> -2 <span class="op">&lt;</span>read2<span class="op">&gt;</span> -o <span class="op">&lt;</span>pair_id<span class="op">&gt;</span></span></code></pre>
</div>
<ol start="4" style="list-style-type: decimal"><li>
<strong>Aggregating Reports with MultiQC</strong>: Finally, the
pipeline employs <a href="https://multiqc.info/" class="external-link">MultiQC</a> to
aggregate logs and outputs from the FastQC and Salmon. MultiQC scans the
outputs and compiles a summary report, which provides an overview of the
results and highlights any areas that may need further
investigation.</li>
</ol><div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> multiqc .</span></code></pre>
</div>
<p>This pipeline provides an overview of RNA-Seq data, from quality
control to expression quantification, culminating in an aggregated
report for easy interpretation of the results.</p>
<p>To start move the episode’s nextflow scripts in the
<code>scripts/rnaseq_pipeline</code> folder to your home directory.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> cp scripts/rnaseq_pipeline/<span class="pp">*</span> .</span></code></pre>
</div>
<p>This folder contains files we will be modifying in this episode.</p>
<section><h2 class="section-heading" id="define-the-pipeline-parameters">Define the pipeline parameters<a class="anchor" aria-label="anchor" href="#define-the-pipeline-parameters"></a></h2>
<hr class="half-width"><p>The first thing we want to do when writing a pipeline is define the
pipeline parameters. The script <code>script1.nf</code> defines the
pipeline input parameters.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co">//script1.nf</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>params<span class="op">.</span>reads <span class="op">=</span> <span class="st">"data/yeast/reads/*_{1,2}.fq.gz"</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>params<span class="op">.</span>transcriptome <span class="op">=</span> <span class="st">"data/yeast/transcriptome/*.fa.gz"</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>println <span class="st">"reads: </span><span class="ss">$params</span><span class="op">.</span><span class="ss">reads</span><span class="st">"</span></span></code></pre>
</div>
<p>Run it by using the following command:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script1.nf</span></code></pre>
</div>
<p>We can specify a different input parameter using the
<code>--&lt;params&gt;</code> option, for example :</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>$ nextflow run script1<span class="op">.</span>nf <span class="op">--</span>reads <span class="st">"data/yeast/reads/ref1*_{1,2}.fq.gz"</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>reads: data/yeast/reads/ref1*_{1,2}.fq.gz</code></pre>
</div>
<div id="add-a-parameter" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="add-a-parameter" class="callout-inner">
<h3 class="callout-title">Add a parameter</h3>
<div class="callout-content">
<p>Modify the <code>script1.nf</code> adding a third parameter named
<code>outdir</code> and set it to <code>results</code>. This parameter
will be used as the pipeline output directory.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>params<span class="op">.</span>outdir <span class="op">=</span> <span class="st">"results"</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>It can be useful to print the pipeline parameters to the screen. This
can be done using the the <code>log.info</code> command and a multiline
string statement. The string method <code>.stripIndent()</code> command
is used to remove the indentation on multi-line strings.
<code>log.info</code> also saves the output to the log execution file
<code>.nextflow.log</code>.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>log<span class="op">.</span><span class="fu">info</span> <span class="st">"""</span><span class="op">\</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="st">         transcriptome: </span><span class="ss">${</span>params<span class="op">.</span>transcriptome<span class="ss">}</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="st">         """</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>         <span class="op">.</span><span class="fu">stripIndent</span><span class="op">()</span></span></code></pre>
</div>
<div id="log.info" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="log.info" class="callout-inner">
<h3 class="callout-title">log.info</h3>
<div class="callout-content">
<p>Modify the <code>script1.nf</code> to print all the pipeline
parameters by using a single <code>log.info</code> command and a
multiline string statement. See an example <a href="https://github.com/nextflow-io/rnaseq-nf/blob/3b5b49f/main.nf#L41-L48" class="external-link">here</a>.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script1.nf</span></code></pre>
</div>
<p>Look at the output log <code>.nextflow.log</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>Below is an example log.info command printing all the pipeline
parameters.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>log<span class="op">.</span><span class="fu">info</span> <span class="st">"""</span><span class="op">\</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="st">        R N A S E Q - N F   P I P E L I N E    </span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="st">        ===================================</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="st">        transcriptome: </span><span class="ss">${</span>params<span class="op">.</span>transcriptome<span class="ss">}</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="st">        reads        : </span><span class="ss">${</span>params<span class="op">.</span>reads<span class="ss">}</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="st">        outdir       : </span><span class="ss">${</span>params<span class="op">.</span>outdir<span class="ss">}</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="st">        """</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>        <span class="op">.</span><span class="fu">stripIndent</span><span class="op">()</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">$</span> less .nextflow.log</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="recap">Recap<a class="anchor" aria-label="anchor" href="#recap"></a></h3>
<p>In this step you have learned:</p>
<ul><li><p>How to define parameters in your pipeline script.</p></li>
<li><p>How to pass parameters by using the command line.</p></li>
<li><p>The use of <code>$var</code> and <code>${var}</code> variable
placeholders.</p></li>
<li><p>How to use multiline strings.</p></li>
<li><p>How to use <code>log.info</code> to print information and save it
in the log execution file.</p></li>
</ul></div>
</section><section><h2 class="section-heading" id="create-a-transcriptome-index-file">Create a transcriptome index file<a class="anchor" aria-label="anchor" href="#create-a-transcriptome-index-file"></a></h2>
<hr class="half-width"><p>Nextflow allows the execution of any command or user script by using
a <code>process</code> definition.</p>
<p>For example,</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">$</span> salmon index <span class="at">--threads</span> <span class="va">$task</span>.cpus <span class="at">-t</span> <span class="va">$transcriptome</span> <span class="at">-i</span> index</span></code></pre>
</div>
<p>A process is defined by providing three main declarations:</p>
<ol style="list-style-type: decimal"><li>The process <a href="https://www.nextflow.io/docs/latest/process.html#inputs" class="external-link">inputs</a>,</li>
<li>The process <a href="https://www.nextflow.io/docs/latest/process.html#outputs" class="external-link">outputs</a>
</li>
<li>Finally the command <a href="https://www.nextflow.io/docs/latest/process.html#script" class="external-link">script</a>.</li>
</ol><p>The second example, <code>script2.nf</code> adds,</p>
<ol style="list-style-type: decimal"><li>The process <code>INDEX</code> which generate a directory with the
index of the transcriptome. This process takes one input, a
transcriptome file, and emits one output a salmon index directory.</li>
<li>A queue Channel <code>transcriptome_ch</code> taking the
transcriptome file defined in params variable
<code>params.transcriptome</code>.</li>
<li>Finally the script adds a <code>workflow</code> definition block
which calls the <code>INDEX</code> process.</li>
</ol><div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co">//script2.nf</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co">/*</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="co"> * pipeline input parameters</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="co"> */</span></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>params<span class="op">.</span>reads <span class="op">=</span> <span class="st">"data/yeast/reads/*_{1,2}.fq.gz"</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a>params<span class="op">.</span>transcriptome <span class="op">=</span> <span class="st">"data/yeast/transcriptome/Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz"</span></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a>params<span class="op">.</span>outdir <span class="op">=</span> <span class="st">"results"</span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a>println <span class="st">"""</span><span class="op">\</span></span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a><span class="st">         R N A S E Q - N F   P I P E L I N E</span></span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a><span class="st">         ===================================</span></span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a><span class="st">         transcriptome: </span><span class="ss">${</span>params<span class="op">.</span>transcriptome<span class="ss">}</span></span>
<span id="cb16-15"><a href="#cb16-15" tabindex="-1"></a><span class="st">         reads        : </span><span class="ss">${</span>params<span class="op">.</span>reads<span class="ss">}</span></span>
<span id="cb16-16"><a href="#cb16-16" tabindex="-1"></a><span class="st">         outdir       : </span><span class="ss">${</span>params<span class="op">.</span>outdir<span class="ss">}</span></span>
<span id="cb16-17"><a href="#cb16-17" tabindex="-1"></a><span class="st">         """</span></span>
<span id="cb16-18"><a href="#cb16-18" tabindex="-1"></a>         <span class="op">.</span><span class="fu">stripIndent</span><span class="op">()</span></span>
<span id="cb16-19"><a href="#cb16-19" tabindex="-1"></a></span>
<span id="cb16-20"><a href="#cb16-20" tabindex="-1"></a></span>
<span id="cb16-21"><a href="#cb16-21" tabindex="-1"></a><span class="st">/*</span></span>
<span id="cb16-22"><a href="#cb16-22" tabindex="-1"></a><span class="st"> * define the `INDEX` process that create a binary index</span></span>
<span id="cb16-23"><a href="#cb16-23" tabindex="-1"></a><span class="st"> * given the transcriptome file</span></span>
<span id="cb16-24"><a href="#cb16-24" tabindex="-1"></a><span class="st"> */</span></span>
<span id="cb16-25"><a href="#cb16-25" tabindex="-1"></a>process INDEX <span class="op">{</span></span>
<span id="cb16-26"><a href="#cb16-26" tabindex="-1"></a></span>
<span id="cb16-27"><a href="#cb16-27" tabindex="-1"></a>    input<span class="op">:</span></span>
<span id="cb16-28"><a href="#cb16-28" tabindex="-1"></a>    path transcriptome</span>
<span id="cb16-29"><a href="#cb16-29" tabindex="-1"></a></span>
<span id="cb16-30"><a href="#cb16-30" tabindex="-1"></a>    output<span class="op">:</span></span>
<span id="cb16-31"><a href="#cb16-31" tabindex="-1"></a>    path <span class="st">'index'</span></span>
<span id="cb16-32"><a href="#cb16-32" tabindex="-1"></a></span>
<span id="cb16-33"><a href="#cb16-33" tabindex="-1"></a>    script<span class="op">:</span></span>
<span id="cb16-34"><a href="#cb16-34" tabindex="-1"></a>    <span class="st">"""</span></span>
<span id="cb16-35"><a href="#cb16-35" tabindex="-1"></a><span class="st">    salmon index --threads </span><span class="ss">$task</span><span class="op">.</span><span class="ss">cpus</span><span class="st"> -t </span><span class="ss">$transcriptome</span><span class="st"> -i index</span></span>
<span id="cb16-36"><a href="#cb16-36" tabindex="-1"></a><span class="st">    """</span></span>
<span id="cb16-37"><a href="#cb16-37" tabindex="-1"></a><span class="op">}</span></span>
<span id="cb16-38"><a href="#cb16-38" tabindex="-1"></a></span>
<span id="cb16-39"><a href="#cb16-39" tabindex="-1"></a>transcriptome_ch <span class="op">=</span> channel<span class="op">.</span><span class="fu">fromPath</span><span class="op">(</span>params<span class="op">.</span>transcriptome<span class="op">)</span></span>
<span id="cb16-40"><a href="#cb16-40" tabindex="-1"></a></span>
<span id="cb16-41"><a href="#cb16-41" tabindex="-1"></a>workflow <span class="op">{</span></span>
<span id="cb16-42"><a href="#cb16-42" tabindex="-1"></a>   <span class="fu">INDEX</span><span class="op">()</span></span>
<span id="cb16-43"><a href="#cb16-43" tabindex="-1"></a><span class="op">}</span></span></code></pre>
</div>
<p>Try to run it by using the command:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script2.nf</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>N E X T F L O W  ~  version 22.04.0
Launching `script2.nf` [happy_brown] DSL2 - revision: 90e932bb8d
R N A S E Q - N F   P I P E L I N E
===================================
transcriptome: data/yeast/transcriptome/Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz
reads        : data/yeast/reads/*_{1,2}.fq.gz
outdir       : results

Process `INDEX` declares 1 input channel but 0 were specified

 -- Check script 'script2.nf' at line: 41 or see '.nextflow.log' file for more details</code></pre>
</div>
<p>The execution will fail because the program the process,
<code>INDEX</code> , has not been passed any input channel.</p>
<p>Add the <code>transcriptome_ch</code> channel to the
<code>INDEX</code> process call.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="op">[</span>truncated<span class="op">]</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>workflow <span class="op">{</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>   <span class="fu">INDEX</span><span class="op">(</span>transcriptome_ch<span class="op">)</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="op">}</span></span></code></pre>
</div>
<p>Now try to run it again by using the command:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script2.nf</span></code></pre>
</div>
<p>Now the workflow will run successfully.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>N E X T F L O W  ~  version 22.04.0
Launching `script2.nf` [mad_aryabhata] DSL2 - revision: 811396b67b
R N A S E Q - N F   P I P E L I N E
===================================
transcriptome:data/yeast/transcriptome/Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz
reads        : data/yeast/reads/*_{1,2}.fq.gz
outdir       : results

executor &gt;  local (1)
[c0/418d78] process &gt; INDEX (1) [100%] 1 of 1 ✔
</code></pre>
</div>
<p>The <code>INDEX</code> process also defines one <code>output</code>
channel. This channel will be populated with <code>index</code>
directory created during process. To view the contents of the channel we
can use the <code>view</code> operator.</p>
<div id="view-the-contents-of-the-index_ch" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="view-the-contents-of-the-index_ch" class="callout-inner">
<h3 class="callout-title">View the contents of the index_ch</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Assign the output of the <code>INDEX</code> process to the variable
<code>index_ch</code>.</li>
<li>View the contents of the <code>index_ch</code> channel by using the
<code>view</code> operator.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="op">[..</span>truncated<span class="op">..]</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>workflow <span class="op">{</span></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>  index_ch<span class="op">=</span><span class="fu">INDEX</span><span class="op">(</span>transcriptome_ch<span class="op">)</span></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>  index_ch<span class="op">.</span><span class="fu">view</span><span class="op">()</span></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a><span class="op">}</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="recap-1">Recap<a class="anchor" aria-label="anchor" href="#recap-1"></a></h3>
<p>In this step you have learned:</p>
<ul><li><p>How to define a process executing a custom command</p></li>
<li><p>How process inputs are declared</p></li>
<li><p>How to assign a channel as input to a process call</p></li>
<li><p>How process outputs are declared</p></li>
<li><p>How to print the content of a channel
<code>view()</code></p></li>
</ul></div>
</section><section><h2 class="section-heading" id="collect-read-files-by-pairs">Collect read files by pairs<a class="anchor" aria-label="anchor" href="#collect-read-files-by-pairs"></a></h2>
<hr class="half-width"><p>This step shows how to match <strong>read</strong> files into pairs,
so they can be mapped by salmon.</p>
<p>The script <code>script3.nf</code> adds a line to create a channel,
<code>read_pairs_ch</code>, containing fastq read pair files using the
<code>fromFilePairs</code> channel factory.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co">//script3.nf</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>nextflow<span class="op">.</span>enable<span class="op">.</span>dsl <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a><span class="co">/*</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a><span class="co"> * pipeline input parameters</span></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a><span class="co"> */</span></span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a>params<span class="op">.</span>reads <span class="op">=</span> <span class="st">"data/yeast/reads/ref1_{1,2}.fq.gz"</span></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a>params<span class="op">.</span>transcriptome <span class="op">=</span> <span class="st">"data/yeast/transcriptome/Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz"</span></span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a>params<span class="op">.</span>outdir <span class="op">=</span> <span class="st">"results"</span></span>
<span id="cb23-10"><a href="#cb23-10" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" tabindex="-1"></a>log<span class="op">.</span><span class="fu">info</span> <span class="st">"""</span><span class="op">\</span></span>
<span id="cb23-12"><a href="#cb23-12" tabindex="-1"></a><span class="st">         R N A S E Q - N F   P I P E L I N E</span></span>
<span id="cb23-13"><a href="#cb23-13" tabindex="-1"></a><span class="st">         ===================================</span></span>
<span id="cb23-14"><a href="#cb23-14" tabindex="-1"></a><span class="st">         transcriptome: </span><span class="ss">${</span>params<span class="op">.</span>transcriptome<span class="ss">}</span></span>
<span id="cb23-15"><a href="#cb23-15" tabindex="-1"></a><span class="st">         reads        : </span><span class="ss">${</span>params<span class="op">.</span>reads<span class="ss">}</span></span>
<span id="cb23-16"><a href="#cb23-16" tabindex="-1"></a><span class="st">         outdir       : </span><span class="ss">${</span>params<span class="op">.</span>outdir<span class="ss">}</span></span>
<span id="cb23-17"><a href="#cb23-17" tabindex="-1"></a><span class="st">         """</span></span>
<span id="cb23-18"><a href="#cb23-18" tabindex="-1"></a>         <span class="op">.</span><span class="fu">stripIndent</span><span class="op">()</span></span>
<span id="cb23-19"><a href="#cb23-19" tabindex="-1"></a></span>
<span id="cb23-20"><a href="#cb23-20" tabindex="-1"></a></span>
<span id="cb23-21"><a href="#cb23-21" tabindex="-1"></a>read_pairs_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromFilePairs</span><span class="op">(</span> params<span class="op">.</span>reads <span class="op">)</span></span></code></pre>
</div>
<p>We can view the contents of the <code>read_pairs_ch</code> by adding
the following statement as the last line:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>read_pairs_ch<span class="op">.</span><span class="fu">view</span><span class="op">()</span></span></code></pre>
</div>
<p>Now if we execute it with the following command:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script3.nf</span></code></pre>
</div>
<p>It will print an output similar to the one shown below that shows how
the <code>read_pairs_ch</code> channel emits a tuple. The tuple is
composed of two elements, where the first is the pattern matched by the
glob pattern <code>data/yeast/reads/ref1_{1,2}.fq.g</code>, defined by
the variable <code>params.reads</code> , and the second is a list
representing the actual files.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[..truncated..]
[ref1, [data/yeast/reads/ref1_1.fq.gz,data/yeast/reads/ref1_2.fq.gz]]</code></pre>
</div>
<p>To read in other read pairs we can specify a different glob pattern
in the <code>params.reads</code> variable by using <code>--reads</code>
options on the command line. For example, the following command would
read in add the ref samples:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script3.nf <span class="at">--reads</span> <span class="st">'data/yeast/reads/ref*_{1,2}.fq.gz'</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[..truncated..]
[ref2, [data/yeast/reads/ref2_1.fq.gz, data/yeast/reads/ref2_2.fq.gz]]
[ref3, [data/yeast/reads/ref3_1.fq.gz, data/yeast/reads/ref3_2.fq.gz]]
[ref1, [data/yeast/reads/ref1_1.fq.gz, data/yeast/reads/ref1_2.fq.gz]]</code></pre>
</div>
<p><strong>Note</strong> File paths including one or more wildcards ie.
<code>*</code>, <code>?</code>, etc. MUST be wrapped in single-quoted
characters to avoid Bash expanding the glob pattern on the command
line.</p>
<p>We can also add a argument, <code>checkIfExists: true</code> , to the
<code>fromFilePairs</code> channel factory to return an message if the
file doesn’t exist.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="co">//script3.nf</span></span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a><span class="op">[..</span>truncated<span class="op">..]</span></span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a>read_pairs_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromFilePairs</span><span class="op">(</span> params<span class="op">.</span>reads<span class="op">,</span> checkIfExists<span class="op">:</span> <span class="kw">true</span> <span class="op">)</span></span></code></pre>
</div>
<p>If we now run the script with the <code>--reads</code> parameter
<code>data/yeast/reads/*_1,2}.fq.gz</code></p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script3.nf <span class="at">--reads</span> <span class="st">'data/yeast/reads/*_1,2}.fq.gz'</span></span></code></pre>
</div>
<p>it will return the message .</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[..truncated..]
No such file: data/yeast/reads/*_1,2}.fq.gz</code></pre>
</div>
<div id="read-in-all-read-pairs" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="read-in-all-read-pairs" class="callout-inner">
<h3 class="callout-title">Read in all read pairs</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Add the <code>checkIfExists: true</code> argument to the
<code>fromFilePairs</code> channel factory in
<code>script3.nf</code>.</li>
<li>Using the command line parameter <code>--reads</code>, add a glob
pattern to read in all the read pairs files from the
<code>data/yeast/reads</code> directory.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a>read_pairs_ch <span class="op">=</span><span class="bu">Channel</span><span class="op">.</span><span class="fu">fromFilePairs</span><span class="op">(</span>params<span class="op">.</span>reads<span class="op">,</span> checkIfExists<span class="op">:</span> <span class="kw">true</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="ex">nextflow</span> run script3.nf <span class="at">--reads</span> <span class="st">'data/yeast/reads/*_{1,2}.fq.gz'</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[..truncated..]
[temp33_1, [data/yeast/reads/temp33_1_1.fq.gz, data/yeast/reads/temp33_1_2.fq.gz]]
[ref2, [data/yeast/reads/ref2_1.fq.gz, data/yeast/reads/ref2_2.fq.gz]]
[temp33_3, [data/yeast/reads/temp33_3_1.fq.gz, data/yeast/reads/temp33_3_2.fq.gz]]
[ref3, [data/yeast/reads/ref3_1.fq.gz, data/yeast/reads/ref3_2.fq.gz]]
[temp33_2, [data/yeast/reads/temp33_2_1.fq.gz,data/yeast/reads/temp33_2_2.fq.gz]]
[etoh60_2, [data/yeast/reads/etoh60_2_1.fq.gz,data/yeast/reads/etoh60_2_2.fq.gz]]
[ref1, [data/yeast/reads/ref1_1.fq.gz, data/yeast/reads/ref1_2.fq.gz]]
[etoh60_3, [data/yeast/reads/etoh60_3_1.fq.gz, data/yeast/reads/etoh60_3_2.fq.gz]]
[etoh60_1, [data/yeast/reads/etoh60_1_1.fq.gz, data/yeast/reads/etoh60_1_2.fq.gz]]</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="recap-2">Recap<a class="anchor" aria-label="anchor" href="#recap-2"></a></h3>
<p>In this step you have learned:</p>
<ul><li><p>How to use <code>fromFilePairs</code> to handle read pair
files</p></li>
<li><p>How to use the <code>checkIfExists</code> option to check input
file existence</p></li>
</ul></div>
</section><section><h2 class="section-heading" id="perform-expression-quantification">Perform expression quantification<a class="anchor" aria-label="anchor" href="#perform-expression-quantification"></a></h2>
<hr class="half-width"><p>The script <code>script4.nf</code>;</p>
<ol style="list-style-type: decimal"><li>Adds the quantification process, <code>QUANT</code>.</li>
<li>Calls the <code>QUANT</code> process in the workflow block.</li>
</ol><div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="co">//script4.nf</span></span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a><span class="op">..</span>truncated<span class="op">..</span></span>
<span id="cb35-3"><a href="#cb35-3" tabindex="-1"></a><span class="st">/*</span></span>
<span id="cb35-4"><a href="#cb35-4" tabindex="-1"></a><span class="st"> * Run Salmon to perform the quantification of expression using</span></span>
<span id="cb35-5"><a href="#cb35-5" tabindex="-1"></a><span class="st"> * the index and the matched read files</span></span>
<span id="cb35-6"><a href="#cb35-6" tabindex="-1"></a><span class="st"> */</span></span>
<span id="cb35-7"><a href="#cb35-7" tabindex="-1"></a>process QUANT <span class="op">{</span></span>
<span id="cb35-8"><a href="#cb35-8" tabindex="-1"></a></span>
<span id="cb35-9"><a href="#cb35-9" tabindex="-1"></a>    input<span class="op">:</span></span>
<span id="cb35-10"><a href="#cb35-10" tabindex="-1"></a>    each index</span>
<span id="cb35-11"><a href="#cb35-11" tabindex="-1"></a>    tuple <span class="fu">val</span><span class="op">(</span>pair_id<span class="op">),</span> <span class="fu">path</span><span class="op">(</span>reads<span class="op">)</span></span>
<span id="cb35-12"><a href="#cb35-12" tabindex="-1"></a></span>
<span id="cb35-13"><a href="#cb35-13" tabindex="-1"></a>    output<span class="op">:</span></span>
<span id="cb35-14"><a href="#cb35-14" tabindex="-1"></a>    <span class="fu">path</span><span class="op">(</span>pair_id<span class="op">)</span></span>
<span id="cb35-15"><a href="#cb35-15" tabindex="-1"></a></span>
<span id="cb35-16"><a href="#cb35-16" tabindex="-1"></a>    script<span class="op">:</span></span>
<span id="cb35-17"><a href="#cb35-17" tabindex="-1"></a>    <span class="st">"""</span></span>
<span id="cb35-18"><a href="#cb35-18" tabindex="-1"></a><span class="st">    salmon quant --threads </span><span class="ss">$task</span><span class="op">.</span><span class="ss">cpus</span><span class="st"> --libType=U -i </span><span class="ss">$index</span><span class="st"> -1 </span><span class="ss">${</span>reads<span class="op">[</span><span class="dv">0</span><span class="op">]</span><span class="ss">}</span><span class="st"> -2 </span><span class="ss">${</span>reads<span class="op">[</span><span class="dv">1</span><span class="op">]</span><span class="ss">}</span><span class="st"> -o </span><span class="ss">$pair_id</span></span>
<span id="cb35-19"><a href="#cb35-19" tabindex="-1"></a><span class="st">    """</span></span>
<span id="cb35-20"><a href="#cb35-20" tabindex="-1"></a><span class="op">}</span></span>
<span id="cb35-21"><a href="#cb35-21" tabindex="-1"></a><span class="op">..</span>truncated<span class="op">..</span></span>
<span id="cb35-22"><a href="#cb35-22" tabindex="-1"></a><span class="fu">workflow</span> <span class="op">{</span></span>
<span id="cb35-23"><a href="#cb35-23" tabindex="-1"></a>  read_pairs_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromFilePairs</span><span class="op">(</span> params<span class="op">.</span>reads<span class="op">,</span> checkIfExists<span class="op">:</span><span class="kw">true</span> <span class="op">)</span></span>
<span id="cb35-24"><a href="#cb35-24" tabindex="-1"></a>  transcriptome_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromPath</span><span class="op">(</span> params<span class="op">.</span>transcriptome<span class="op">,</span> checkIfExists<span class="op">:</span><span class="kw">true</span> <span class="op">)</span></span>
<span id="cb35-25"><a href="#cb35-25" tabindex="-1"></a></span>
<span id="cb35-26"><a href="#cb35-26" tabindex="-1"></a>  index_ch<span class="op">=</span><span class="fu">INDEX</span><span class="op">(</span>transcriptome_ch<span class="op">)</span></span>
<span id="cb35-27"><a href="#cb35-27" tabindex="-1"></a>  quant_ch<span class="op">=</span><span class="fu">QUANT</span><span class="op">(</span>index_ch<span class="op">,</span>read_pairs_ch<span class="op">)</span></span>
<span id="cb35-28"><a href="#cb35-28" tabindex="-1"></a><span class="op">}</span></span></code></pre>
</div>
<p>The <code>index_ch</code> channel, declared as output in the
<code>INDEX</code> process, is used as the first input argument to the
<code>QUANT</code> process.</p>
<p>The second input argument of the <code>QUANT</code> process, the
<code>read_pairs_ch</code> channel, is a tuple composed of two elements:
the <code>pair_id</code> and the <code>reads</code>.</p>
<p>Execute it by using the following command:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script4.nf</span></code></pre>
</div>
<p>You will see the execution of the index and quantification
process.</p>
<p>Re run the command using the <code>-resume</code> option</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script4.nf <span class="at">-resume</span></span></code></pre>
</div>
<p>The <code>-resume</code> option cause the execution of any step that
has been already processed to be skipped.</p>
<p>Try to execute it with more read files as shown below:</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script4.nf <span class="at">-resume</span> <span class="at">--reads</span> <span class="st">'data/yeast/reads/ref*_{1,2}.fq.gz'</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>N E X T F L O W  ~  version 21.04.0
Launching `script4.nf` [shrivelled_brenner] - revision: c21df6839e
R N A S E Q - N F   P I P E L I N E
===================================
transcriptome: data/yeast/transcriptome/Saccharomyces_c
erevisiae.R64-1-1.cdna.all.fa.gz

reads        : data/yeast/reads/ref*_{1,2}.fq.gz
outdir       : results

executor &gt;  local (8)
[02/3742cf] process &gt; INDEX     [100%] 1 of 1, cached: 1 ✔
[9a/be3483] process &gt; QUANT (9) [100%] 3 of 3, cached: 1 ✔</code></pre>
</div>
<p>You will notice that the <code>INDEX</code> step and one of the
<code>QUANT</code> steps has been cached, and the quantification process
is executed more than one time.</p>
<p>When your input channel contains multiple data items Nextflow, where
possible, parallelises the execution of your pipeline.</p>
<p>In these situations it is useful to add a <code>tag</code> directive
to add some descriptive text to instance of the process being run.</p>
<div id="add-a-tag-directive" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="add-a-tag-directive" class="callout-inner">
<h3 class="callout-title">Add a tag directive</h3>
<div class="callout-content">
<p>Add a <code>tag</code> directive to the <code>QUANT</code> process of
<code>script4.nf</code> to provide a more readable execution log.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a>tag <span class="st">"quantification on </span><span class="ss">$pair_id</span><span class="st">"</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Data produced by the workflow during a process will be saved in the
working directory, by default a directory named <code>work</code>. The
working directory should be considered a temporary storage space and any
data you wish to save at the end of the workflow should be specified in
the process output with the final storage location defined in the
<code>publishDir</code> directive.</p>
<p><strong>Note:</strong> by default the <code>publishDir</code>
directive creates a symbolic link to the files in the working this
behaviour can be changed using the <code>mode</code> parameter.</p>
</section><section><h2 class="section-heading" id="add-a-publishdir-directive">Add a publishDir directive<a class="anchor" aria-label="anchor" href="#add-a-publishdir-directive"></a></h2>
<hr class="half-width"><p>Add a <code>publishDir</code> directive to the quantification process
of <code>script4.nf</code> to store the process results into folder
specified by the <code>params.outdir</code> Nextflow variable. Include
the <code>publishDir</code> <code>mode</code> option to copy the
output.</p>
<div id="challenge6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a>publishDir <span class="st">"</span><span class="ss">${</span>params<span class="op">.</span>outdir<span class="ss">}</span><span class="st">/quant"</span><span class="op">,</span> mode<span class="op">:</span><span class="st">'copy'</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="recap-3">Recap<a class="anchor" aria-label="anchor" href="#recap-3"></a></h3>
<p>In this step you have learned:</p>
<ul><li><p>How to connect two processes by using the channel
declarations.</p></li>
<li><p>How to resume the script execution skipping already already
computed steps.</p></li>
<li><p>How to use the <code>tag</code> directive to provide a more
readable execution output.</p></li>
<li><p>How to use the <code>publishDir</code> to store a process results
in a path of your choice.</p></li>
</ul></div>
</section><section><h2 class="section-heading" id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a></h2>
<hr class="half-width"><p>This step implements a quality control step for your input reads. The
input to the <code>FASTQC</code> process is the same
<code>read_pairs_ch</code> that is provided as input to the
quantification process <code>QUANT</code> .</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="co">//script5.nf</span></span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a><span class="op">[..</span>truncated<span class="op">..]</span></span>
<span id="cb42-3"><a href="#cb42-3" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" tabindex="-1"></a><span class="st">/*</span></span>
<span id="cb42-5"><a href="#cb42-5" tabindex="-1"></a><span class="st"> * Run fastQC to check quality of reads files</span></span>
<span id="cb42-6"><a href="#cb42-6" tabindex="-1"></a><span class="st"> */</span></span>
<span id="cb42-7"><a href="#cb42-7" tabindex="-1"></a>process FASTQC <span class="op">{</span></span>
<span id="cb42-8"><a href="#cb42-8" tabindex="-1"></a>    tag <span class="st">"FASTQC on </span><span class="ss">$sample_id</span><span class="st">"</span></span>
<span id="cb42-9"><a href="#cb42-9" tabindex="-1"></a>    cpus <span class="dv">1</span></span>
<span id="cb42-10"><a href="#cb42-10" tabindex="-1"></a></span>
<span id="cb42-11"><a href="#cb42-11" tabindex="-1"></a>    input<span class="op">:</span></span>
<span id="cb42-12"><a href="#cb42-12" tabindex="-1"></a>    tuple <span class="fu">val</span><span class="op">(</span>sample_id<span class="op">),</span> <span class="fu">path</span><span class="op">(</span>reads<span class="op">)</span></span>
<span id="cb42-13"><a href="#cb42-13" tabindex="-1"></a></span>
<span id="cb42-14"><a href="#cb42-14" tabindex="-1"></a>    output<span class="op">:</span></span>
<span id="cb42-15"><a href="#cb42-15" tabindex="-1"></a>    <span class="fu">path</span><span class="op">(</span><span class="st">"fastqc_</span><span class="ss">${</span>sample_id<span class="ss">}</span><span class="st">_logs"</span><span class="op">)</span></span>
<span id="cb42-16"><a href="#cb42-16" tabindex="-1"></a></span>
<span id="cb42-17"><a href="#cb42-17" tabindex="-1"></a>    script<span class="op">:</span></span>
<span id="cb42-18"><a href="#cb42-18" tabindex="-1"></a>    <span class="st">"""</span></span>
<span id="cb42-19"><a href="#cb42-19" tabindex="-1"></a><span class="st">    mkdir fastqc_</span><span class="ss">${</span>sample_id<span class="ss">}</span><span class="st">_logs</span></span>
<span id="cb42-20"><a href="#cb42-20" tabindex="-1"></a><span class="st">    fastqc -o fastqc_</span><span class="ss">${</span>sample_id<span class="ss">}</span><span class="st">_logs -f fastq -q </span><span class="ss">${</span>reads<span class="ss">}</span><span class="st"> -t </span><span class="ss">${</span>task<span class="op">.</span>cpus<span class="ss">}</span></span>
<span id="cb42-21"><a href="#cb42-21" tabindex="-1"></a><span class="st">    """</span></span>
<span id="cb42-22"><a href="#cb42-22" tabindex="-1"></a><span class="op">}</span></span>
<span id="cb42-23"><a href="#cb42-23" tabindex="-1"></a></span>
<span id="cb42-24"><a href="#cb42-24" tabindex="-1"></a><span class="op">[..</span>truncated<span class="op">..]</span></span>
<span id="cb42-25"><a href="#cb42-25" tabindex="-1"></a></span>
<span id="cb42-26"><a href="#cb42-26" tabindex="-1"></a>workflow <span class="op">{</span></span>
<span id="cb42-27"><a href="#cb42-27" tabindex="-1"></a>  read_pairs_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromFilePairs</span><span class="op">(</span> params<span class="op">.</span>reads<span class="op">,</span> checkIfExists<span class="op">:</span><span class="kw">true</span> <span class="op">)</span></span>
<span id="cb42-28"><a href="#cb42-28" tabindex="-1"></a>  transcriptome_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromPath</span><span class="op">(</span> params<span class="op">.</span>transcriptome<span class="op">,</span> checkIfExists<span class="op">:</span><span class="kw">true</span> <span class="op">)</span></span>
<span id="cb42-29"><a href="#cb42-29" tabindex="-1"></a></span>
<span id="cb42-30"><a href="#cb42-30" tabindex="-1"></a>  index_ch<span class="op">=</span><span class="fu">INDEX</span><span class="op">(</span>transcriptome_ch<span class="op">)</span></span>
<span id="cb42-31"><a href="#cb42-31" tabindex="-1"></a>  quant_ch<span class="op">=</span><span class="fu">QUANT</span><span class="op">(</span>index_ch<span class="op">,</span>read_pairs_ch<span class="op">)</span></span>
<span id="cb42-32"><a href="#cb42-32" tabindex="-1"></a><span class="op">}</span></span></code></pre>
</div>
<p>Run the script <code>script5.nf</code> by using the following
command:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script5.nf <span class="at">-resume</span></span></code></pre>
</div>
<p>The <code>FASTQC</code> process will not run as the process has not
been declared in the workflow scope.</p>
<div id="add-fastqc-process" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="add-fastqc-process" class="callout-inner">
<h3 class="callout-title">Add FASTQC process</h3>
<div class="callout-content">
<p>Add the <code>FASTQC</code> process to the
<code>workflow scope</code> of <code>script5.nf</code> adding the
<code>read_pairs_ch</code> channel as an input. Run the nextflow script
using the <code>-resume</code> option.</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script5.nf <span class="at">-resume</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a>workflow <span class="op">{</span></span>
<span id="cb45-2"><a href="#cb45-2" tabindex="-1"></a>read_pairs_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromFilePairs</span><span class="op">(</span> params<span class="op">.</span>reads<span class="op">,</span> checkIfExists<span class="op">:</span><span class="kw">true</span> <span class="op">)</span></span>
<span id="cb45-3"><a href="#cb45-3" tabindex="-1"></a>transcriptome_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromPath</span><span class="op">(</span> params<span class="op">.</span>transcriptome<span class="op">,</span> checkIfExists<span class="op">:</span><span class="kw">true</span> <span class="op">)</span></span>
<span id="cb45-4"><a href="#cb45-4" tabindex="-1"></a></span>
<span id="cb45-5"><a href="#cb45-5" tabindex="-1"></a>index_ch <span class="op">=</span> <span class="fu">INDEX</span><span class="op">(</span> transcriptome_ch <span class="op">)</span></span>
<span id="cb45-6"><a href="#cb45-6" tabindex="-1"></a>quant_ch<span class="op">=</span><span class="fu">QUANT</span><span class="op">(</span>index_ch<span class="op">,</span>read_pairs_ch<span class="op">)</span></span>
<span id="cb45-7"><a href="#cb45-7" tabindex="-1"></a>fastqc_ch<span class="op">=</span><span class="fu">FASTQC</span><span class="op">(</span>read_pairs_ch<span class="op">)</span></span></code></pre>
</div>
<p>}</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="recap-4">Recap<a class="anchor" aria-label="anchor" href="#recap-4"></a></h3>
<p>In this step you have learned:</p>
<ul><li>How to use the add a <code>process</code> to the
<code>workflow</code> scope.</li>
<li>Add a channel as input to a <code>process</code>.</li>
</ul></div>
</section><section><h2 class="section-heading" id="multiqc-report">MultiQC report<a class="anchor" aria-label="anchor" href="#multiqc-report"></a></h2>
<hr class="half-width"><p>This step collect the outputs from the quantification and fastqc
steps to create a final report by using the <a href="https://multiqc.info/" class="external-link">MultiQC</a> tool.</p>
<p>The input for the <code>MULTIQC</code> process requires all data in a
single channel element. Therefore, we will need to combine the
<code>FASTQC</code> and <code>QUANT</code> outputs using:</p>
<ul><li>The combining operator <code>mix</code> : combines the items in the
two channels into a single channel</li>
</ul><div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="co">//example of the mix operator</span></span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a>ch1 <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">of</span><span class="op">(</span><span class="dv">1</span><span class="op">,</span><span class="dv">2</span><span class="op">)</span></span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a>ch2 <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">of</span><span class="op">(</span><span class="st">'a'</span><span class="op">)</span></span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a>ch1<span class="op">.</span><span class="fu">mix</span><span class="op">(</span>ch2<span class="op">).</span><span class="fu">view</span><span class="op">()</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fl">1</span></span>
<span><span class="fl">2</span></span>
<span><span class="va">a</span></span></code></pre>
</div>
<ul><li>The transformation operator <code>collect</code> collects all the
items in the new combined channel into a single item.</li>
</ul><div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="co">//example of the collect operator</span></span>
<span id="cb48-2"><a href="#cb48-2" tabindex="-1"></a>ch1 <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">of</span><span class="op">(</span><span class="dv">1</span><span class="op">,</span><span class="dv">2</span><span class="op">,</span><span class="dv">3</span><span class="op">)</span></span>
<span id="cb48-3"><a href="#cb48-3" tabindex="-1"></a>ch1<span class="op">.</span><span class="fu">collect</span><span class="op">().</span><span class="fu">view</span><span class="op">()</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1, 2, 3]</code></pre>
</div>
<div id="combining-operators" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="combining-operators" class="callout-inner">
<h3 class="callout-title">Combining operators</h3>
<div class="callout-content">
<p>Which is the correct way to combined <code>mix</code> and
<code>collect</code> operators so that you have a single channel with
one List item?</p>
<ol style="list-style-type: decimal"><li><code>quant_ch.mix(fastqc_ch).collect()</code></li>
<li><code>quant_ch.collect(fastqc_ch).mix()</code></li>
<li><code>fastqc_ch.mix(quant_ch).collect()</code></li>
<li><code>fastqc_ch.collect(quant_ch).mix()</code></li>
</ol></div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<p>You need to use the <code>mix</code> operator first to combine the
channels followed by the <code>collect</code> operator to collect all
the items in a single item.</p>
</div>
</div>
</div>
</div>
<p>In <code>script6.nf</code> we use the statement
<code>quant_ch.mix(fastqc_ch).collect()</code> to combine and collect
the outputs of the <code>QUANT</code> and <code>FASTQC</code> process to
create the required input for the <code>MULTIQC</code> process.</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="op">[..</span>truncated<span class="op">..]</span></span>
<span id="cb50-2"><a href="#cb50-2" tabindex="-1"></a><span class="co">//script6.nf</span></span>
<span id="cb50-3"><a href="#cb50-3" tabindex="-1"></a><span class="co">/*</span></span>
<span id="cb50-4"><a href="#cb50-4" tabindex="-1"></a><span class="co"> * Create a report using multiQC for the quantification</span></span>
<span id="cb50-5"><a href="#cb50-5" tabindex="-1"></a><span class="co"> * and fastqc processes</span></span>
<span id="cb50-6"><a href="#cb50-6" tabindex="-1"></a><span class="co"> */</span></span>
<span id="cb50-7"><a href="#cb50-7" tabindex="-1"></a>process MULTIQC <span class="op">{</span></span>
<span id="cb50-8"><a href="#cb50-8" tabindex="-1"></a>    publishDir <span class="st">"</span><span class="ss">${</span>params<span class="op">.</span>outdir<span class="ss">}</span><span class="st">/multiqc"</span><span class="op">,</span> mode<span class="op">:</span><span class="st">'copy'</span></span>
<span id="cb50-9"><a href="#cb50-9" tabindex="-1"></a></span>
<span id="cb50-10"><a href="#cb50-10" tabindex="-1"></a>    input<span class="op">:</span></span>
<span id="cb50-11"><a href="#cb50-11" tabindex="-1"></a>    <span class="fu">path</span><span class="op">(</span><span class="st">'*'</span><span class="op">)</span></span>
<span id="cb50-12"><a href="#cb50-12" tabindex="-1"></a></span>
<span id="cb50-13"><a href="#cb50-13" tabindex="-1"></a>    output<span class="op">:</span></span>
<span id="cb50-14"><a href="#cb50-14" tabindex="-1"></a>    <span class="fu">path</span><span class="op">(</span><span class="st">'multiqc_report.html'</span><span class="op">)</span></span>
<span id="cb50-15"><a href="#cb50-15" tabindex="-1"></a></span>
<span id="cb50-16"><a href="#cb50-16" tabindex="-1"></a>    script<span class="op">:</span></span>
<span id="cb50-17"><a href="#cb50-17" tabindex="-1"></a>    <span class="st">"""</span></span>
<span id="cb50-18"><a href="#cb50-18" tabindex="-1"></a><span class="st">    multiqc .</span></span>
<span id="cb50-19"><a href="#cb50-19" tabindex="-1"></a><span class="st">    """</span></span>
<span id="cb50-20"><a href="#cb50-20" tabindex="-1"></a><span class="op">}</span></span>
<span id="cb50-21"><a href="#cb50-21" tabindex="-1"></a></span>
<span id="cb50-22"><a href="#cb50-22" tabindex="-1"></a></span>
<span id="cb50-23"><a href="#cb50-23" tabindex="-1"></a>workflow <span class="op">{</span></span>
<span id="cb50-24"><a href="#cb50-24" tabindex="-1"></a>  read_pairs_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromFilePairs</span><span class="op">(</span> params<span class="op">.</span>reads<span class="op">,</span> checkIfExists<span class="op">:</span><span class="kw">true</span> <span class="op">)</span></span>
<span id="cb50-25"><a href="#cb50-25" tabindex="-1"></a>  transcriptome_ch <span class="op">=</span> <span class="bu">Channel</span><span class="op">.</span><span class="fu">fromPath</span><span class="op">(</span> params<span class="op">.</span>transcriptome<span class="op">,</span> checkIfExists<span class="op">:</span><span class="kw">true</span> <span class="op">)</span></span>
<span id="cb50-26"><a href="#cb50-26" tabindex="-1"></a></span>
<span id="cb50-27"><a href="#cb50-27" tabindex="-1"></a>  index_ch<span class="op">=</span><span class="fu">INDEX</span><span class="op">(</span>transcriptome_ch<span class="op">)</span></span>
<span id="cb50-28"><a href="#cb50-28" tabindex="-1"></a>  quant_ch<span class="op">=</span><span class="fu">QUANT</span><span class="op">(</span>index_ch<span class="op">,</span>read_pairs_ch<span class="op">)</span></span>
<span id="cb50-29"><a href="#cb50-29" tabindex="-1"></a>  fastqc_ch<span class="op">=</span><span class="fu">FASTQC</span><span class="op">(</span>read_pairs_ch<span class="op">)</span></span>
<span id="cb50-30"><a href="#cb50-30" tabindex="-1"></a>  <span class="fu">MULTIQC</span><span class="op">(</span>quant_ch<span class="op">.</span><span class="fu">mix</span><span class="op">(</span>fastqc_ch<span class="op">).</span><span class="fu">collect</span><span class="op">())</span></span>
<span id="cb50-31"><a href="#cb50-31" tabindex="-1"></a><span class="op">}</span></span></code></pre>
</div>
<p>Execute the script with the following command:</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script6.nf <span class="at">--reads</span> <span class="st">'data/yeast/reads/*_{1,2}.fq.gz'</span> <span class="at">-resume</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>N E X T F L O W  ~  version 21.04.0
Launching `script6.nf` [small_franklin] - revision: 9062818659
R N A S E Q - N F   P I P E L I N E
===================================
transcriptome: data/yeast/transcriptome/Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz
reads        : data/yeast/reads/*_{1,2}.fq.gz
outdir       : results

executor &gt;  local (9)
[02/3742cf] process &gt; INDEX                              [100%] 1 of 1, cached: 1 ✔
[9a/be3483] process &gt; QUANT (quantification on etoh60_1) [100%] 9 of 9, cached: 9 ✔
[1f/b7b30a] process &gt; FASTQC (FASTQC on etoh60_1)        [100%] 9 of 9, cached: 1 ✔
[2c/206fef] process &gt; MULTIQC                            [100%] 1 of 1 ✔</code></pre>
</div>
<p>It creates the final report in the results folder in the
<code>${params.outdir}/multiqc</code> directory.</p>
<div class="section level3">
<h3 id="recap-5">Recap<a class="anchor" aria-label="anchor" href="#recap-5"></a></h3>
<p>In this step you have learned:</p>
<ul><li><p>How to collect many outputs to a single input with the
<code>collect</code> operator</p></li>
<li><p>How to mix two channels in a single channel using the
<code>mix</code> operator.</p></li>
<li><p>How to chain two or more operators togethers using the
<code>.</code> operator.</p></li>
</ul></div>
</section><section><h2 class="section-heading" id="handle-completion-event">Handle completion event<a class="anchor" aria-label="anchor" href="#handle-completion-event"></a></h2>
<hr class="half-width"><p>This step shows how to execute an action when the pipeline completes
the execution.</p>
<p><strong>Note:</strong> that Nextflow processes define the execution
of asynchronous tasks i.e. they are not executed one after another as
they are written in the pipeline script as it would happen in a common
imperative programming language.</p>
<p>The script <code>script7..nf</code> uses the
<code>workflow.onComplete</code> event handler to print a confirmation
message when the script completes.</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">GROOVY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode groovy" tabindex="0"><code class="sourceCode groovy"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a>workflow<span class="op">.</span><span class="fu">onComplete</span> <span class="op">{</span></span>
<span id="cb53-2"><a href="#cb53-2" tabindex="-1"></a>	log<span class="op">.</span><span class="fu">info</span> <span class="op">(</span> workflow<span class="op">.</span>success <span class="op">?</span> <span class="st">"</span><span class="sc">\n</span><span class="st">Done! Open the following report in your browser --&gt; </span><span class="ss">$params</span><span class="op">.</span><span class="ss">outdir</span><span class="st">/multiqc/multiqc_report.html</span><span class="sc">\n</span><span class="st">"</span> <span class="op">:</span> <span class="st">"Oops .. something went wrong"</span> <span class="op">)</span></span>
<span id="cb53-3"><a href="#cb53-3" tabindex="-1"></a><span class="op">}</span></span></code></pre>
</div>
<p>This code uses the ternary operator that is a shortcut expression
that is equivalent to an if/else branch assigning some value to a
variable.</p>
<pre class="source"><code>If expression is true? "set value to a" : "else set value to b"</code></pre>
<p>Try to run it by using the following command:</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script7.nf <span class="at">-resume</span> <span class="at">--reads</span> <span class="st">'data/yeast/reads/*_{1,2}.fq.gz'</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[..truncated..]
Done! Open the following report in your browser --&gt; results/multiqc/multiqc_report.html</code></pre>
</div>
</section><section><h2 class="section-heading" id="metrics-and-reports">Metrics and reports<a class="anchor" aria-label="anchor" href="#metrics-and-reports"></a></h2>
<hr class="half-width"><p>Nextflow is able to produce multiple reports and charts providing
several runtime metrics and execution information.</p>
<ul><li><p>The <code>-with-report</code> option enables the creation of the
workflow execution report.</p></li>
<li><p>The <code>-with-trace</code> option enables the create of a tab
separated file containing runtime information for each executed task,
including: submission time, start time, completion time, cpu and memory
used..</p></li>
<li><p>The <code>-with-timeline</code> option enables the creation of
the workflow timeline report showing how processes where executed along
time. This may be useful to identify most time consuming tasks and
bottlenecks. See an example at this <a href="https://www.nextflow.io/docs/latest/tracing.html#timeline-report" class="external-link">link</a>.</p></li>
<li><p>The <code>-with-dag</code> option enables to rendering of the
workflow execution direct acyclic graph representation.
<strong>Note:</strong> this feature requires the installation of <a href="https://graphviz.org/" class="external-link">Graphviz</a>, an open source graph
visualization software, in your system.</p></li>
</ul><p>More information can be found <a href="https://www.nextflow.io/docs/latest/tracing.html" class="external-link">here</a>.</p>
<div id="metrics-and-reports-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="metrics-and-reports-1" class="callout-inner">
<h3 class="callout-title">Metrics and reports</h3>
<div class="callout-content">
<p>Run the script7.nf with the reporting options as shown below:</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb57-1"><a href="#cb57-1" tabindex="-1"></a><span class="ex">$</span> nextflow run script7.nf <span class="at">-resume</span> <span class="at">-with-report</span> <span class="at">-with-trace</span> <span class="at">-with-timeline</span> <span class="at">-with-dag</span> dag.png</span></code></pre>
</div>
<ol style="list-style-type: decimal"><li>Open the file <code>report.html</code> with a browser to see the
report created with the above command.</li>
<li>Check the content of the file <code>trace.txt</code> or view
<code>timeline.html</code> to find the longest running process.</li>
<li>View the dag.png</li>
</ol></div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<p>The <code>INDEX</code> process should be the longest running process.
dag.png <img src="../fig/dag.png" alt="dag" class="figure"> The vertices in the graph
represent the pipeline’s processes and operators, while the edges
represent the data connections (i.e. channels) between them.</p>
</div>
</div>
</div>
</div>
<div id="short-running-tasks" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="short-running-tasks" class="callout-inner">
<h3 class="callout-title">short running tasks</h3>
<div class="callout-content">
<p>Note: runtime metrics may be incomplete for run short running
tasks..</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Nextflow can combined tasks (processes) and manage data flows using
channels into a single pipeline/workflow.</li>
<li>A Workflow can be parameterise using <code>params</code> . These
value of the parameters can be captured in a log file using
<code>log.info</code>
</li>
<li>Nextflow can handle a workflow’s software requirements using several
technologies including the <code>conda</code> package and enviroment
manager.</li>
<li>Workflow steps are connected via their <code>inputs</code> and
<code>outputs</code> using <code>Channels</code>.</li>
<li>Intermediate pipeline results can be transformed using Channel
<code>operators</code> such as <code>combine</code>.</li>
<li>Nextflow can execute an action when the pipeline completes the
execution using the <code>workflow.onComplete</code> event handler to
print a confirmation message.</li>
<li>Nextflow is able to produce multiple reports and charts providing
several runtime metrics and execution information using the command line
options <code>-with-report</code>, <code>-with-trace</code>,
<code>-with-timeline</code> and produce a graph using
<code>-with-dag</code>.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/10-configuration.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/12-nfcore.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/10-configuration.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Nextflow
        </a>
        <a class="chapter-link float-end" href="../instructor/12-nfcore.html" rel="next">
          Next: Deploying nf-core...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/workflows-nextflow/edit/main/episodes/11-Simple_Rna-Seq_pipeline.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/workflows-nextflow/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/workflows-nextflow/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/workflows-nextflow/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:graeme.grimes@igmm.ed.ac.uk&amp;amp;quot;">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.7" class="external-link">pegboard (0.7.7)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/workflows-nextflow/instructor/11-Simple_Rna-Seq_pipeline.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Simple RNA-Seq pipeline",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/workflows-nextflow/instructor/11-Simple_Rna-Seq_pipeline.html",
  "identifier": "https://carpentries-incubator.github.io/workflows-nextflow/instructor/11-Simple_Rna-Seq_pipeline.html",
  "dateCreated": "2021-02-01",
  "dateModified": "2025-01-06",
  "datePublished": "2025-01-21"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

